trl/trainer/srpo_trainer.py

```diff
@@ -0,0 +1,1756 @@
+# SRPO Authors: Eugene Choi, Arash Ahmadian, Matthieu Geist, Olivier Pietquin, Mohammad Gheshlaghi 2024
```

Comment:
Much of this is based heavily on the DPO trainer. That said it felt different enough that it was worth splitting out. Perhaps we could try to inherit from it but I am a bit wary to rely upon inheritance where it can be avoided.